### DISCUSSION TITLE
How to get tiddler's real file path in the disk in nodejs wiki?

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#6189] [Q&A] [answered]
I found drag&drop a large ( 10M ) PDF file will make this PDF unable to display in the wiki.

<img width="1295" alt="截屏2021-11-10 下午11 04 00" src="https://user-images.githubusercontent.com/3746270/141137691-71cf9476-ff93-4ec9-b125-bda0ffc078e3.png">

So I'm gonna write a plugin to provide a button to open this PDF in the system's default image viewer. To do so, I need to get the real path of the tiddler, but `$tw.wiki.getTiddler` doesn't seem to provide this info.

So, is there an API to get the tiddler file path? Or do I need to search this file using glob by myself?

**Comment by joshuafontany** [comment, marked as answer]
On Node, this is found in the $tw.boot.files object, an index-by-title of fileInfo objects. This is not accessible from the browser.

**Reply by linonetwo** [reply]
Thanks, I can read that in tiddlygit, and use IPC to reply to the query from the wiki, so this works for me.

**Comment by Jermolene** [comment]
Hi @linonetwo browsers now limit the size of data URIs which breaks large embedded PDFs. One workaround is to use the static file serving capability - see https://tiddlywiki.com/#Using%20the%20integrated%20static%20file%20server

**Reply by linonetwo** [reply]
Thanks, can I use `canonical_uri` instead?

Because after I drag&drop file into the wiki page, it will be slightly annoying if I still need to open the underlying folder structure and move the file into `./files`

I hope I can change this behavior to automatically add canonical_uri.

So it will lazy load the PDF and embed it into the page.

And that canonical_uri may be a relative path from the `tiddlers` folder, so I can still add a button to open this file in an external app (by asking tiddlygit electron to call system API using canonical_uri path.)

**Reply by saqimtiaz** [reply]
You could use the FileUploads plugin along with the PUT uploader and the node-files-PUT support plugins, all available via my plugin library: https://talk.tiddlywiki.org/t/file-uploads-plugin-updates/461/11?u=saqimtiaz

I think the PUT uploader creates absolute URLs but I have a more feature rich version that I can clean up and publish in the near future if it would be useful, which allows greater control over the canonical URI that is generated.

**Reply by linonetwo** [reply]
Ok, Saq, I will take a look. I just want to allow users to know less about underlying details, and can switch from notion to tiddlywiki easily.

**Reply by linonetwo** [reply]
Oh, I see, @saqimtiaz , nodejs relative path uploader is currently not supported. Looking forward to the update!
